### PR TITLE
Add cluster refresh command for upgrades

### PR DESCRIPTION
**Pull request by gnuoy**
## Possible issues

* The code assumes the Juju terraform provider does not support upgrading channels. When this feature gap is fixed the code can be simplified. 
 
## What's new 
                                                                                                                                                                   
This PR adds the ability to refresh/upgrade the sunbeam deployment. The two
modes of operation are:
                                              
```bash                                                                                     
sunbeam cluster refresh                                                                     
```                                         
                                              
This will upgrade all charm to the latest revision in the current track.

```bash                          
sunbeam cluster refresh --upgrade-release
```                                                                                         
                                                                                            
This will switch all charms to the new channel. The snap now has an sunbeam.versions file that contains all the application channels for the core charms. When a refresh is run with the **--upgrade-release** option then each application channel is checked against the snaps version file and updated if needed.

                                                                                            
## Structure                                                                                
                                                                                            
When a refresh is requested a coordinator class is selected to handle
the upgrade. The coordinator class contains the plan for the upgrade,
the plan contains each step. For a typical upgrade between channels
the might look like: 

```python 
    plan = [
        ValidationCheck(self.jhelper, self.tfhelper),      
        LatestInChannel(self.jhelper),                                                      
        UpgradeControlPlane2023_2(                                                          
            self.jhelper, self.tfhelper, self.channel, "openstack"
        ),                                   
        UpgradeMachineCharms2023_2(                                                         
            self.jhelper, self.tfhelper, self.channel, "controller"
        ),
        UpgradePlugins(self.jhelper, self.tfhelper, self.channel),
    ]
```

In this plan the following will happen:

1. Check if the upgrade is possible
2. Upgrade all charms to the latest version in the current channel
3. Upgrade the control plane charms to the new channel
4. Upgrade machine charms to the new channel
5. Upgrade plugins

## Refresh without channel switch

If a refresh is requested without a channel switch then the
**LatestInChannelCoordinator** class is called. The plan for
this coordinator has a single step **LatestInChannel**. This
will iterate over all the charms in the deployment and see
if a newer version is available in the current channel. If it
is then the charm is upgraded.

## Refresh with channel switch

If a refresh is requested with a channel switch then the **ChannelUpgradeCoordinator** is used.

## Plugins

The **UpgradePlugins** looks after calling the upgrade hook
in the plugins. For plugins derived from the **OpenStackControlPlanePlugin**
this is achieved by running the UpgradeApplicationStep step
plugin class. To support refreshes between channels using the the UpgradeApplicationStep a plugin needs to define the  channels and corresponding tfvars property (if there is one) for each application. This is done using the new **k8s_application_data** method in the plugin class. For example:

```
    @property
    def k8s_application_data(self):
        return {
            "designate": {
                "channel": OPENSTACK_CHANNEL,
                "tfvars_channel_var": None,
            },
            "bind": {
                "channel": "9/edge",
                "tfvars_channel_var": None,
            },
        }
```

### CODE REVIEW

File: sunbeam-python/sunbeam/versions.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+CERT_AUTH_CHANNEL = "latest/beta"
+
+# The lists of services are needed for switching charm channels outside
+# of the terraform provider. If it ok to upgrade in one big-bang and
+# the juju terraform provider supports it then the upgrades can be
+# done by simply updating the tfvars and these lists are not needed.
+OPENSTACK_SERVICES_K8S = {
+    "cinder-ceph": OPENSTACK_CHANNEL,
+    "cinder": OPENSTACK_CHANNEL,
+    "glance": "2023.2/candidate",
```
Comment (gboutry) [inline sunbeam-python/sunbeam/versions.py:33]: Is this a workaround while we fix the bug in glance releases on charmhub ?

File: sunbeam-python/sunbeam/commands/upgrades/inter_channel.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        :jhelper: Helper for interacting with pylibjuju
+        :tfhelper: Helper for interaction with Terraform
+        :model: Name of model containing charms.
+        """
+        super().__init__(name, description)
+        self.jhelper = jhelper
+        self.tfhelper = tfhelper
+        self.model = model
+
+    def get_upgrade_strategy(self) -> List[Dict[str, Union[Callable, List]]]:
```
Comment (gboutry) [inline sunbeam-python/sunbeam/commands/upgrades/inter_channel.py:61]: This return type is hard to read, the doc in the function is great, but to simplify things, maybe introducing a typed dict might make things simpler ?

```python
from typing import TypedDict, Callable
class UpgradeStrategy(TypedDict):
    upgrade_f: Callable[[list[str], str], None]
    applications: list[str]

def get_upgrade_strategy(self) -> list[UpgradeStrategy]:
```

File: sunbeam-python/sunbeam/commands/upgrades/inter_channel.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+    def post_upgrade_tasks(self) -> None:
+        """Tasks to run after the upgrade."""
+        return
+
+    def upgrade_applications(
+        self,
+        application_list: List[str],
+        model: str,
+        expect_wls: Optional[Dict[str, str]] = None,
```
Comment (gboutry) [inline sunbeam-python/sunbeam/commands/upgrades/inter_channel.py:100]:
```suggestion
        expect_wls: Optional[Dict[str, list[str]]] = None,
```

File: sunbeam-python/sunbeam/commands/upgrades/inter_channel.py (hunk 2 above)
Comment (gboutry) [reply sunbeam-python/sunbeam/commands/upgrades/inter_channel.py:61]: There are other occurrences, like sunbeam-python/sunbeam/plugins/interface/v1/openstack.py L288 and L303

File: sunbeam-python/sunbeam/commands/juju.py [hunk 4]
```diff
@@ -171,6 +172,104 @@ def add_cloud(self, cloud_type: str, cloud_name: str) -> bool:
...
+            for d in available_charm_data["channels"][track][risk]
+            if arch in d["architectures"]
+        ]
+        assert len(channel_data) == 1, "Unexpected candidate charms {}".format(
+            len(channel_data)
+        )
+        return int(channel_data[0]["revision"])
+
+    def revision_update_needed(
+        self, application_name: str, model: str, status: Union[dict, None] = None
```
[the remark this answers is not in the document]
Comment (gnuoy) [reply sunbeam-python/sunbeam/commands/juju.py]: Done

File: sunbeam-python/sunbeam/versions.py (hunk 1 above)
Comment (gnuoy) [reply sunbeam-python/sunbeam/versions.py]: yes it was, I've removed it now

File: sunbeam-python/sunbeam/commands/juju.py [hunk 5]
```diff
@@ -171,6 +172,104 @@ def add_cloud(self, cloud_type: str, cloud_name: str) -> bool:
...
+
+    def normalise_channel(self, channel: str) -> str:
+        """Expand channel if it is using abbreviation.
+
+        Juju supports abbreviating latest/{risk} to {risk}. This expands it.
+
+        :param channel: Channel string to normalise
+        """
+        if channel in ["stable", "candidate", "beta", "edge"]:
+            channel = f"latest/{channel}"
```
[the remark this answers is not in the document]
Comment (gnuoy) [reply sunbeam-python/sunbeam/commands/juju.py:212]: This is to deal with charms. Specifically the microk8s charm:

```
$ juju status microceph | grep -E '^App|\smicroceph.*edge'
App                   Version  Status  Scale  Charm                 Channel        Rev  Exposed  Message
microceph                      active      1  microceph             edge             9  no       

$ juju info microceph
name: microceph
publisher: OpenStack Charmers
summary: microceph
description: |
  MicroCeph is a lightweight way of deploying a Ceph cluster designed
  to provide excellent performance, reliability, and scalability.
store-url: https://charmhub.io/microceph
charm-id: BcMT3WaOBBrvzPEVK05CCsugg7iGIG00
supports: ubuntu@22.04
subordinate: false
relations:
  provides:
    ceph: ceph-client
  requires: {}
channels: |
  latest/stable:     –
  latest/candidate:  –
  latest/beta:       –
  latest/edge:       9  2023-08-11  (9)  17MB  amd64  ubuntu@22.04

```

File: sunbeam-python/sunbeam/commands/juju.py [hunk 6]
```diff
@@ -171,6 +172,104 @@ def add_cloud(self, cloud_type: str, cloud_name: str) -> bool:
 
         return True
 
+    def get_available_charm_revision(
```
[the remark this answers is not in the document]
Comment (gnuoy) [reply sunbeam-python/sunbeam/commands/juju.py]: Only reason was I didn't know about it! Fixed now

**Review by javacruft — APPROVED** [review]

**Pull request merged**